types/bluebird/bluebird-tests.ts

```diff
@@ -373,6 +373,30 @@ fooProm = fooProm.catch(CustomError, reason => {
 	fooProm = fooProm.catch(CustomError1, CustomError2, CustomError3, error => {});
 	fooProm = fooProm.catch(CustomError1, CustomError2, CustomError3, CustomError4, error => {});
 	fooProm = fooProm.catch(CustomError1, CustomError2, CustomError3, CustomError4, CustomError5, error => {});
+
+	const booPredicate1 = (error: CustomError1) => true;
+	const booPredicate2 = (error: [number]) => true;
+	const booPredicate3 = (error: string) => true;
+	const booPredicate4 = (error: Object) => true;
+	const booPredicate5 = (error: any) => true;
+
+	fooProm = fooProm.catch(booPredicate1, error => {});
+	fooProm = fooProm.catch(booPredicate1, booPredicate2, error => {});
+	fooProm = fooProm.catch(booPredicate1, booPredicate2, booPredicate3, error => {});
+	fooProm = fooProm.catch(booPredicate1, booPredicate2, booPredicate3, booPredicate4, error => {});
+	fooProm = fooProm.catch(booPredicate1, booPredicate2, booPredicate3, booPredicate4, booPredicate5, error => {});
+
+	const booObject1 = new CustomError1();
+	const booObject2 = [400, 500];
+	const booObject3 = ["Error1", "Error2"];
+	const booObject4 = {code: 400};
+	const booObject5: any = null;
+
+	fooProm = fooProm.catch(booObject1, error => {});
+	fooProm = fooProm.catch(booObject1, booObject2, error => {});
+	fooProm = fooProm.catch(booObject1, booObject2, booObject3, error => {});
+	fooProm = fooProm.catch(booObject1, booObject2, booObject3, booObject4, error => {});
+	fooProm = fooProm.catch(booObject1, booObject2, booObject3, booObject4, booObject5, error => {});
```

Comment:
Hi @jcready, thanks for your review.

Indeed, actually there are two cases covered here:

- For *any* `catch` where *all* `E extends Error`

  ```typescript
  filter1: (new (...args: any[]) => E1)
  ```

- For *any* `catch` where *all* _generic_  `E` are _predicates_ or _object_-like

  ```typescript
  filter1: ((error: E1) => boolean) | (E1 & object)
  ```

And both are mutually exclusive

```typescript
fooProm.catch(
  CustomError1,
  booObject2,
  booPredicate3,
  error => {}
);
```

So, your example will try to match the 2nd case, and I think it will pass because even if `CustomError1` is not an _instance_ it is still an _object_-like.